### PR TITLE
Implement post-contest statistics tab

### LINKED ISSUE #706: Post-contest statistics tab
Have been requested a large number of times. [We have the technology](https://dmoj.ca/stats/language/), so we can just fill in some graphs post-contest under a Stats tab.

Some contests-specific stats could include:

* [ ] stacked bar chart for status codes on each problem
* [ ] bar chart for full AC %
* [ ] score distribution
* [ ] top submission languages
* [ ] top language AC rates
* [ ] ...

### PR DESCRIPTION
**Pull request by Ninjaclasher**
Closes #706.

### CODE REVIEW

File: templates/stats/media-js.html [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+                    position: 'right',
+                    labels: {
+                        fontColor: 'black',
+                        boxWidth: 20,
+                    },
+                },
+            },
+        });
+    }
+    function draw_bar_chart(data, $chart) {
```
Comment (Xyene) [inline templates/stats/media-js.html]: newline

File: templates/stats/media-js.html [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                    callbacks: {
+                        label: function(tooltipItem, data) {
+                            return (Math.round(tooltipItem.value * 100) / 100) + '%';
+                        },
+                    },
+                },
+            },
+        });
+    }
+    function draw_stacked_bar_chart(data, $chart) {
```
Comment (Xyene) [inline templates/stats/media-js.html]: newline

File: judge/views/submission.py [hunk 3]
```diff
@@ -252,6 +253,7 @@ def get_context_data(self, **kwargs):
         context['selected_statuses'] = self.selected_statuses
 
         context['results_json'] = mark_safe(json.dumps(self.get_result_data()))
+        context['results_colours'] = mark_safe(json.dumps(settings.DMOJ_STATS_SUBMISSION_RESULT_COLOURS))
```
Comment (Xyene) [inline judge/views/submission.py]: `results_colors_json`

File: judge/views/contests.py [hunk 4]
```diff
@@ -485,6 +490,71 @@ def render(self):
...
+            for category in _get_result_data(defaultdict(int, status_counts[i]))['categories']:
+                result_data[category['code']][i] = category['count']
+
+        stats = {
+            'problem_status_count': {
+                'labels': labels,
+                'datasets': [
+                    {
+                        'label': name,
+                        'backgroundColor': settings.DMOJ_STATS_SUBMISSION_RESULT_COLOURS[name],
```
Comment (Xyene) [inline judge/views/contests.py]: `COLOR`

**Comment by Ninjaclasher** [comment]
The requested changes have been made in 9ef7ce5.

File: judge/views/contests.py [hunk 5]
```diff
@@ -485,6 +490,71 @@ def render(self):
...
+class ContestStats(TitleMixin, ContestMixin, DetailView):
+    template_name = 'contest/stats.html'
+
+    def get_title(self):
+        return _('%s Statistics') % self.object.name
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+
+        if not self.object.ended:
```
Comment (Xyene) [inline judge/views/contests.py]: I'm not sure about how sensitive this data is. It's already fully public for non-hidden-scoreboard contests, so maybe it should be visible before the end of the contest? (At least, it should be visible if `self.object.is_editable_by(...)`)

Everything else looks fine, btw.

Comment (Ninjaclasher) [reply judge/views/contests.py]: The contest statistics page exposes problem names, which are hidden from users who are not in the contest. A commit has been added to allow contest organizers to see the statistics before the end of the contest.

**Pull request merged**